File: wear/src/main/java/com/thewizrd/simplewear/controls/ActionButtonViewModel.java
package com.thewizrd.simplewear.controls;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;

import androidx.annotation.ColorRes;
import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import com.thewizrd.shared_resources.helpers.Action;
import com.thewizrd.shared_resources.helpers.Actions;
import com.thewizrd.shared_resources.helpers.DNDChoice;
import com.thewizrd.shared_resources.helpers.MultiChoiceAction;
import com.thewizrd.shared_resources.helpers.NormalAction;
import com.thewizrd.shared_resources.helpers.RingerChoice;
import com.thewizrd.shared_resources.helpers.ToggleAction;
import com.thewizrd.shared_resources.helpers.ValueAction;
import com.thewizrd.shared_resources.utils.JSONParser;
import com.thewizrd.simplewear.App;
import com.thewizrd.simplewear.MusicPlayerActivity;
import com.thewizrd.simplewear.R;
import com.thewizrd.simplewear.ValueActionActivity;
import com.thewizrd.simplewear.WearableListenerActivity;

public class ActionButtonViewModel {
    private Action action;
    private @DrawableRes
    int mDrawableID;
    private @ColorRes
    int mButtonBackgroundColor;
    private String mActionLabel;
    private String mStateLabel;

    public Action getAction() {
        return action;
    }

    @DrawableRes
    public int getDrawableID() {
        return mDrawableID;
    }

    @ColorRes
    public int getButtonBackgroundColor() {
        return mButtonBackgroundColor;
    }

    public Actions getActionType() {
        return action.getAction();
    }

    public String getActionLabel() {
        return mActionLabel;
    }

    public String getStateLabel() {
        return mStateLabel;
    }

    public ActionButtonViewModel(@NonNull Action action) {
        this.action = action;
        mDrawableID = R.drawable.ic_cc_clear;
        mButtonBackgroundColor = R.color.buttonDisabled;
        initialize(action);
    }

    private void initialize(Action action) {
        mButtonBackgroundColor = R.color.colorPrimary;
        mDrawableID = R.drawable.ic_cc_clear;

        if (action instanceof ToggleAction) {
            ToggleAction tA = (ToggleAction) action;

            if (!tA.isActionSuccessful()) {
                // Revert state
                tA.setEnabled(!tA.isEnabled());
            }

            mButtonBackgroundColor = tA.isEnabled() ? R.color.buttonEnabled : R.color.buttonDisabled;
            updateIconAndLabel();
        } else if (action instanceof MultiChoiceAction) {
            MultiChoiceAction mA = (MultiChoiceAction) action;

            if (!mA.isActionSuccessful()) {
                // Revert state
                mA.setChoice(mA.getChoice() - 1);
            }

            updateIconAndLabel();
        } else if (action != null) {
            updateIconAndLabel();
        } else {
            throw new IllegalArgumentException("Action class is invalid!!");
        }
    }

    public void onClick(Activity activityContext) {
        action.setActionSuccessful(true);

        if (action instanceof ValueAction) {
            Intent intent = new Intent(activityContext, ValueActionActivity.class)
                    .putExtra(ValueActionActivity.EXTRA_ACTION, getActionType());
            activityContext.startActivityForResult(intent, -1);
        } else if (action instanceof NormalAction && action.getAction() == Actions.MUSICPLAYBACK) {
            Intent intent = new Intent(activityContext, MusicPlayerActivity.class);
            activityContext.startActivityForResult(intent, -1);
        } else {
            if (action instanceof ToggleAction) {
                ToggleAction tA = (ToggleAction) action;
                tA.setEnabled(!tA.isEnabled());
                mButtonBackgroundColor = R.color.colorPrimaryDark;
            } else if (action instanceof MultiChoiceAction) {
                MultiChoiceAction mA = (MultiChoiceAction) action;
                int currentChoice = mA.getChoice();
                int newChoice = currentChoice + 1;
                mA.setChoice(newChoice);
                updateIconAndLabel();
            }

            LocalBroadcastManager.getInstance(activityContext)
                    .sendBroadcast(new Intent(WearableListenerActivity.ACTION_CHANGED)
                            .putExtra(WearableListenerActivity.EXTRA_ACTIONDATA,
                                    JSONParser.serializer(action, Action.class)));
        }
    }

    private void updateIconAndLabel() {
        ToggleAction tA;
        MultiChoiceAction mA;
        Context context = App.getInstance().getAppContext();

        switch (getActionType()) {
            case WIFI:
                tA = (ToggleAction) action;
                mDrawableID = tA.isEnabled() ? R.drawable.ic_network_wifi_white_24dp : R.drawable.ic_signal_wifi_off_white_24dp;
                mActionLabel = context.getString(R.string.action_wifi);
                mStateLabel = tA.isEnabled() ? context.getString(R.string.state_on) : context.getString(R.string.state_off);
                break;
            case BLUETOOTH:
                tA = (ToggleAction) action;
                mDrawableID = tA.isEnabled() ? R.drawable.ic_bluetooth_white_24dp : R.drawable.ic_bluetooth_disabled_white_24dp;
                mActionLabel = context.getString(R.string.action_bt);
                mStateLabel = tA.isEnabled() ? context.getString(R.string.state_on) : context.getString(R.string.state_off);
                break;
            case MOBILEDATA:
                tA = (ToggleAction) action;
                mDrawableID = tA.isEnabled() ? R.drawable.ic_network_cell_white_24dp : R.drawable.ic_signal_cellular_off_white_24dp;
                mActionLabel = context.getString(R.string.action_mobiledata);
                mStateLabel = tA.isEnabled() ? context.getString(R.string.state_on) : context.getString(R.string.state_off);
                break;
            case LOCATION:
                tA = (ToggleAction) action;
                mDrawableID = tA.isEnabled() ? R.drawable.ic_location_on_white_24dp : R.drawable.ic_location_off_white_24dp;
                mActionLabel = context.getString(R.string.action_location);
                mStateLabel = tA.isEnabled() ? context.getString(R.string.state_on) : context.getString(R.string.state_off);
                break;
            case TORCH:
                tA = (ToggleAction) action;
                mDrawableID = R.drawable.ic_lightbulb_outline_white_24dp;
                mActionLabel = context.getString(R.string.action_torch);
                mStateLabel = tA.isEnabled() ? context.getString(R.string.state_on) : context.getString(R.string.state_off);
                break;
            case LOCKSCREEN:
                mDrawableID = R.drawable.ic_lock_outline_white_24dp;
                mActionLabel = context.getString(R.string.action_lockscreen);
                mStateLabel = null;
                break;
            case VOLUME:
                mDrawableID = R.drawable.ic_volume_up_white_24dp;
                mActionLabel = context.getString(R.string.action_volume);
                mStateLabel = null;
                break;
            case DONOTDISTURB:
                mA = (MultiChoiceAction) action;

                mActionLabel = context.getString(R.string.action_dnd);

                DNDChoice dndChoice = DNDChoice.valueOf(mA.getChoice());
                switch (dndChoice) {
                    case OFF:
                        mDrawableID = R.drawable.ic_do_not_disturb_off_white_24dp;
                        mStateLabel = context.getString(R.string.dndstate_off);
                        break;
                    case PRIORITY:
                        mDrawableID = R.drawable.ic_error_white_24dp;
                        mStateLabel = context.getString(R.string.dndstate_priority);
                        break;
                    case ALARMS:
                        mDrawableID = R.drawable.ic_alarm_white_24dp;
                        mStateLabel = context.getString(R.string.dndstate_alarms);
                        break;
                    case SILENCE:
                        mDrawableID = R.drawable.ic_notifications_off_white_24dp;
                        mStateLabel = context.getString(R.string.dndstate_silence);
                        break;
                }
                break;
            case RINGER:
                mA = (MultiChoiceAction) action;

                mActionLabel = context.getString(R.string.action_ringer);

                RingerChoice ringerChoice = RingerChoice.valueOf(mA.getChoice());
                switch (ringerChoice) {
                    case VIBRATION:
                        mDrawableID = R.drawable.ic_vibration_white_24dp;
                        mStateLabel = context.getString(R.string.ringerstate_vib);
                        break;
                    case SOUND:
                        mDrawableID = R.drawable.ic_notifications_active_white_24dp;
                        mStateLabel = context.getString(R.string.ringerstate_sound);
                        break;
                    case SILENT:
                        mDrawableID = R.drawable.ic_volume_off_white_24dp;
                        mStateLabel = context.getString(R.string.ringerstate_silent);
                        break;
                }
                break;
            case MUSICPLAYBACK:
                mDrawableID = R.drawable.ic_play_circle_filled_white_24dp;
                mActionLabel = context.getString(R.string.action_musicplayback);
                mStateLabel = null;
                break;
        }
    }
}

File: wear/src/main/java/com/thewizrd/simplewear/controls/ActionButtonViewModel.kt
package com.thewizrd.simplewear.controls

import android.os.Build
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import androidx.navigation.NavController
import androidx.recyclerview.widget.DiffUtil
import com.thewizrd.shared_resources.actions.Action
import com.thewizrd.shared_resources.actions.ActionStatus
import com.thewizrd.shared_resources.actions.Actions
import com.thewizrd.shared_resources.actions.DNDChoice
import com.thewizrd.shared_resources.actions.LocationState
import com.thewizrd.shared_resources.actions.MultiChoiceAction
import com.thewizrd.shared_resources.actions.NormalAction
import com.thewizrd.shared_resources.actions.RingerChoice
import com.thewizrd.shared_resources.actions.ToggleAction
import com.thewizrd.shared_resources.actions.ValueAction
import com.thewizrd.shared_resources.actions.ValueDirection
import com.thewizrd.shared_resources.sleeptimer.SleepTimerHelper
import com.thewizrd.simplewear.R
import com.thewizrd.simplewear.ui.navigation.Screen
import java.util.Objects

class ActionButtonViewModel(val action: Action) {
    @get:DrawableRes
    @DrawableRes
    var drawableResId: Int
        private set

    val actionType: Actions
        get() = action.actionType

    @StringRes
    var actionLabelResId: Int = 0
        private set

    @StringRes
    var stateLabelResId: Int = 0
        private set

    var buttonState: Boolean? = null
        private set

    companion object {
        val DIFF_CALLBACK = object : DiffUtil.ItemCallback<ActionButtonViewModel>() {
            override fun areItemsTheSame(
                oldItem: ActionButtonViewModel,
                newItem: ActionButtonViewModel
            ): Boolean {
                return oldItem.actionType == newItem.actionType
            }

            override fun areContentsTheSame(
                oldItem: ActionButtonViewModel,
                newItem: ActionButtonViewModel
            ): Boolean {
                return Objects.equals(oldItem.action, newItem.action)
            }
        }

        fun getViewModelFromAction(action: Actions): ActionButtonViewModel {
            return when (action) {
                Actions.WIFI,
                Actions.BLUETOOTH,
                Actions.MOBILEDATA,
                Actions.TORCH,
                Actions.HOTSPOT ->
                    ActionButtonViewModel(ToggleAction(action, true))
                Actions.LOCATION ->
                    ActionButtonViewModel(
                        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.P)
                            MultiChoiceAction(action)
                        else
                            ToggleAction(action, true)
                    )
                Actions.LOCKSCREEN,
                Actions.MUSICPLAYBACK,
                Actions.SLEEPTIMER,
                Actions.APPS,
                Actions.PHONE,
                Actions.GESTURES ->
                    ActionButtonViewModel(NormalAction(action))
                Actions.VOLUME, Actions.BRIGHTNESS ->
                    ActionButtonViewModel(ValueAction(action, ValueDirection.UP))
                Actions.DONOTDISTURB ->
                    ActionButtonViewModel(
                        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.P)
                            MultiChoiceAction(action)
                        else
                            ToggleAction(action, true)
                    )
                Actions.RINGER ->
                    ActionButtonViewModel(MultiChoiceAction(action))
            }
        }
    }

    init {
        drawableResId = R.drawable.ic_close_white_24dp
        buttonState = false
        initialize(action)
    }

    private fun initialize(action: Action?) {
        buttonState = true
        drawableResId = R.drawable.ic_close_white_24dp

        if (action is ToggleAction) {
            val tA = action

            if (!tA.isActionSuccessful) {
                // Revert state
                tA.isEnabled = !tA.isEnabled
            }

            buttonState = tA.isEnabled
            updateIconAndLabel()
        } else if (action is MultiChoiceAction) {
            val mA = action

            if (!mA.isActionSuccessful) {
                // Revert state
                mA.choice = mA.choice - 1
            }

            buttonState = mA.choice > 0
            updateIconAndLabel()
        } else if (action != null) {
            updateIconAndLabel()
        } else {
            throw IllegalArgumentException("Action class is invalid!!")
        }
    }

    fun onClick(
        navController: NavController,
        onActionChanged: (Action) -> Unit,
        onActionStatus: (Action) -> Unit
    ) {
        action.isActionSuccessful = true

        if (action is ValueAction) {
            navController.navigate("${Screen.ValueAction.route}/${actionType.value}")
        } else if (action is NormalAction && action.actionType == Actions.MUSICPLAYBACK) {
            navController.navigate(Screen.MediaPlayerList.route)
        } else if (action is NormalAction && action.actionType == Actions.SLEEPTIMER) {
            if (SleepTimerHelper.isSleepTimerInstalled()) {
                SleepTimerHelper.launchSleepTimer()
            } else {
                action.setActionSuccessful(ActionStatus.PERMISSION_DENIED)

                onActionStatus.invoke(action)
            }
        } else if (action is NormalAction && action.actionType == Actions.APPS) {
            navController.navigate(Screen.AppLauncher.route)
        } else if (action is NormalAction && action.actionType == Actions.PHONE) {
            navController.navigate(Screen.CallManager.route)
        } else if (action is NormalAction && action.actionType == Actions.GESTURES) {
            navController.navigate(Screen.GesturesAction.route)
        } else {
            if (action is ToggleAction) {
                val tA = action
                tA.isEnabled = !tA.isEnabled
                buttonState = null
            } else if (action is MultiChoiceAction) {
                val mA = action
                val currentChoice = mA.choice
                val newChoice = currentChoice + 1
                mA.choice = newChoice
                updateIconAndLabel()
            }

            onActionChanged.invoke(action)
        }
    }

    private fun updateIconAndLabel() {
        val tA: ToggleAction
        val mA: MultiChoiceAction

        when (actionType) {
            Actions.WIFI -> {
                tA = action as ToggleAction
                drawableResId =
                    if (tA.isEnabled) R.drawable.ic_network_wifi_white_24dp else R.drawable.ic_signal_wifi_off_white_24dp
                actionLabelResId = R.string.action_wifi
                stateLabelResId = if (tA.isEnabled) R.string.state_on else R.string.state_off
            }
            Actions.BLUETOOTH -> {
                tA = action as ToggleAction
                drawableResId =
                    if (tA.isEnabled) R.drawable.ic_bluetooth_white_24dp else R.drawable.ic_bluetooth_disabled_white_24dp
                actionLabelResId = R.string.action_bt
                stateLabelResId = if (tA.isEnabled) R.string.state_on else R.string.state_off
            }
            Actions.MOBILEDATA -> {
                tA = action as ToggleAction
                drawableResId =
                    if (tA.isEnabled) R.drawable.ic_network_cell_white_24dp else R.drawable.ic_signal_cellular_off_white_24dp
                actionLabelResId = R.string.action_mobiledata
                stateLabelResId = if (tA.isEnabled) R.string.state_on else R.string.state_off
            }
            Actions.LOCATION -> {
                actionLabelResId = R.string.action_location

                val locationState = if (action is ToggleAction) {
                    if (action.isEnabled) LocationState.HIGH_ACCURACY else LocationState.OFF
                } else {
                    mA = action as MultiChoiceAction
                    LocationState.valueOf(mA.choice)
                }
                when (locationState) {
                    LocationState.OFF -> {
                        drawableResId = R.drawable.ic_location_off_white_24dp
                        stateLabelResId = R.string.state_off
                    }
                    LocationState.SENSORS_ONLY -> {
                        drawableResId = R.drawable.ic_baseline_gps_fixed_24dp
                        stateLabelResId = R.string.locationstate_sensorsonly
                    }
                    LocationState.BATTERY_SAVING -> {
                        drawableResId = R.drawable.ic_outline_location_on_24dp
                        stateLabelResId = R.string.locationstate_batterysaving
                    }
                    LocationState.HIGH_ACCURACY -> {
                        drawableResId = R.drawable.ic_location_on_white_24dp
                        stateLabelResId = if (action is ToggleAction) {
                            R.string.state_on
                        } else {
                            R.string.locationstate_highaccuracy
                        }
                    }
                }
            }
            Actions.TORCH -> {
                tA = action as ToggleAction
                drawableResId = R.drawable.ic_lightbulb_outline_white_24dp
                actionLabelResId = R.string.action_torch
                stateLabelResId = if (tA.isEnabled) R.string.state_on else R.string.state_off
            }
            Actions.LOCKSCREEN -> {
                drawableResId = R.drawable.ic_lock_outline_white_24dp
                actionLabelResId = R.string.action_lockscreen
                stateLabelResId = 0
            }
            Actions.VOLUME -> {
                drawableResId = R.drawable.ic_volume_up_white_24dp
                actionLabelResId = R.string.action_volume
                stateLabelResId = 0
            }
            Actions.DONOTDISTURB -> {
                actionLabelResId = R.string.action_dnd

                val dndChoice = if (action is ToggleAction) {
                    if (action.isEnabled) DNDChoice.PRIORITY else DNDChoice.OFF
                } else {
                    mA = action as MultiChoiceAction
                    DNDChoice.valueOf(mA.choice)
                }
                when (dndChoice) {
                    DNDChoice.OFF -> {
                        drawableResId = R.drawable.ic_do_not_disturb_off_white_24dp
                        stateLabelResId = R.string.state_off
                    }
                    DNDChoice.PRIORITY -> {
                        drawableResId = R.drawable.ic_error_white_24dp
                        stateLabelResId = if (Build.VERSION.SDK_INT < Build.VERSION_CODES.P) {
                            R.string.dndstate_priority
                        } else {
                            R.string.state_on
                        }
                    }
                    DNDChoice.ALARMS -> {
                        drawableResId = R.drawable.ic_alarm_white_24dp
                        stateLabelResId = R.string.dndstate_alarms
                    }
                    DNDChoice.SILENCE -> {
                        drawableResId = R.drawable.ic_notifications_off_white_24dp
                        stateLabelResId = R.string.dndstate_silence
                    }
                }
            }
            Actions.RINGER -> {
                mA = action as MultiChoiceAction

                actionLabelResId = R.string.action_ringer

                when (RingerChoice.valueOf(mA.choice)) {
                    RingerChoice.VIBRATION -> {
                        drawableResId = R.drawable.ic_vibration_white_24dp
                        stateLabelResId = R.string.ringerstate_vib
                    }
                    RingerChoice.SOUND -> {
                        drawableResId = R.drawable.ic_notifications_active_white_24dp
                        stateLabelResId = R.string.ringerstate_sound
                    }
                    RingerChoice.SILENT -> {
                        drawableResId = R.drawable.ic_volume_off_white_24dp
                        stateLabelResId = R.string.ringerstate_silent
                    }
                }
            }
            Actions.MUSICPLAYBACK -> {
                drawableResId = R.drawable.ic_play_circle_filled_white_24dp
                actionLabelResId = R.string.action_musicplayback
                stateLabelResId = 0
            }
            Actions.SLEEPTIMER -> {
                drawableResId = R.drawable.ic_sleep_timer
                actionLabelResId = R.string.action_sleeptimer
                stateLabelResId = 0
            }
            Actions.APPS -> {
                drawableResId = R.drawable.ic_apps_white_24dp
                actionLabelResId = R.string.action_apps
                stateLabelResId = 0
            }
            Actions.PHONE -> {
                drawableResId = R.drawable.ic_phone_24dp
                actionLabelResId = R.string.action_phone
                stateLabelResId = 0
            }
            Actions.BRIGHTNESS -> {
                drawableResId = R.drawable.ic_brightness_medium
                actionLabelResId = R.string.action_brightness
                stateLabelResId = 0
            }
            Actions.HOTSPOT -> {
                tA = action as ToggleAction
                drawableResId = R.drawable.ic_wifi_tethering
                actionLabelResId = R.string.action_hotspot
                stateLabelResId =
                    if (tA.isEnabled) R.string.state_on else R.string.state_off
            }
            Actions.GESTURES -> {
                drawableResId = R.drawable.ic_touch_app
                actionLabelResId = R.string.action_gestures
                stateLabelResId = 0
            }
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ActionButtonViewModel

        if (action != other.action) return false
        if (actionType != other.actionType) return false
        if (actionLabelResId != other.actionLabelResId) return false
        if (stateLabelResId != other.stateLabelResId) return false
        if (buttonState != other.buttonState) return false

        return true
    }

    override fun hashCode(): Int {
        var result = action.hashCode()
        result = 31 * result + actionType.hashCode()
        result = 31 * result + actionLabelResId
        result = 31 * result + stateLabelResId
        result = 31 * result + (buttonState?.hashCode() ?: 0)
        return result
    }
}